server/src/main/java/org/apache/cassandra/sidecar/acl/authorization/CachedAuthorizationHandler.java

```diff
@@ -0,0 +1,157 @@
+public class CachedAuthorizationHandler extends AuthorizationHandlerImpl
+{
+    private static final HttpException FORBIDDEN_EXCEPTION = new HttpException(403);
+    private final AccessControlConfiguration accessControlConfiguration;
+    private final AuthorizationParameterValidateHandler authZParameterValidateHandler;
+    private final AdminIdentityResolver adminIdentityResolver;
+    private final AuthMetrics authMetrics;
+    private final CacheStatsCounter cacheMetrics;
+    private final Cache<AuthorizationCacheKey, Boolean> authorizationCache;
+    // This is overridden since Vert.x does not expose this
+    private BiConsumer<RoutingContext, AuthorizationContext> variableHandler;
+
+    public CachedAuthorizationHandler(AccessControlConfiguration accessControlConfiguration,
+                                      AuthorizationParameterValidateHandler authZParameterValidateHandler,
+                                      AdminIdentityResolver adminIdentityResolver,
+                                      Authorization authorization,
+                                      SidecarMetrics sidecarMetrics)
+    {
+        super(authorization);
+        this.accessControlConfiguration = accessControlConfiguration;
+        this.authZParameterValidateHandler = authZParameterValidateHandler;
+        this.adminIdentityResolver = adminIdentityResolver;
+        this.authMetrics = sidecarMetrics.server().auth();
+        this.cacheMetrics = sidecarMetrics.server().cache().authorizationCacheMetrics;
+        this.authorizationCache = initCache();
+    }
+
+    @Override
+    public void handle(RoutingContext ctx)
+    {
+        long startTimeNanos = System.nanoTime();
+        authZParameterValidateHandler.handle(ctx);
+        if (ctx.failed()) // failed due to validation
+        {
+            return;
+        }
+
+        User user = ctx.user();
+        AuthorizationContext authorizationContext = AuthorizationContext.create(user);
+        if (this.variableHandler != null)
+        {
+            this.variableHandler.accept(ctx, authorizationContext);
+        }
+
+        AtomicBoolean ctxNextCalled = new AtomicBoolean(false);
+
+        AuthorizationCacheKey key = AuthorizationCacheKey.create(authorizationContext);
+        Boolean authorized = authorizationCache.get(key, k -> {
+            List<String> identities = extractIdentities(user);
+
+            // Admin identities bypass route specific authorization checks
+            if (!identities.isEmpty() && identities.stream().anyMatch(adminIdentityResolver::isAdmin))
+            {
+                return true;
+            }
+
+            super.handle(ctx);
+            if (!ctx.failed())
+            {
+                ctxNextCalled.set(true);
+                long durationNanos = System.nanoTime() - startTimeNanos;
+                authMetrics.authorizationTime.metric.update(durationNanos, TimeUnit.NANOSECONDS);
+                return true;
+            }
+            return false;
```

Comment:
I am concerned about the complexity in this mapping function. It is supposed to be simple and short.

Reply:
Switched to AsyncCache as suggested